test/remote/gateways/remote_sage_pay_test.rb

```diff
@@ -97,9 +107,75 @@ def setup
       phone: '0161 123 4567'
     }
 
+    @options_v4 = {
+      billing_address: {
+        name: 'Tekin Suleyman',
+        address1: 'Flat 10 Lapwing Court',
+        address2: 'West Didsbury',
+        city: 'Manchester',
+        county: 'Greater Manchester',
+        country: 'GB',
+        zip: 'M20 2PS'
+      },
+      shipping_address: {
+        name: 'Tekin Suleyman',
+        address1: '120 Grosvenor St',
+        city: 'Manchester',
+        county: 'Greater Manchester',
+        country: 'GB',
+        zip: 'M1 7QW'
+      },
+      order_id: generate_unique_id,
+      description: 'Store purchase',
+      ip: '86.150.65.37',
+      email: 'tekin@tekin.co.uk',
+      phone: '0161 123 4567',
+      protocol_version: '4.00',
+      three_ds_2: {
+        channel: 'browser',
+        browser_info: {
+          accept_header: 'unknown',
+          depth: 48,
+          java: true,
+          language: 'US',
+          height: 1000,
+          width: 500,
+          timezone: '-120',
+          user_agent: 'unknown',
+          browser_size: '05'
+        },
+        notification_url: 'https://example.com/notification'
+      }
+    }
+
     @amount = 100
   end
 
+  # Protocol 4
+  def test_successful_purchase_v4
+    assert response = @gateway.purchase(@amount, @mastercard, @options_v4)
+    assert_success response
+
+    assert response.test?
+    assert !response.authorization.blank?
+  end
+
+  def test_three_ds_challenge_purchase_v4
+    assert response = @gateway.purchase(@amount, @mastercard, @options_v4.merge(apply_3d_secure: 1))
```

Comment:
just a question of my own curiosity - does modifying `@options_v4` with the `.merge` change the object for all future tests, or does it get reset outside of this test automagically?

Reply:
`.merge` combines 2 hashes and returns a new hash while `.merge!` combines the provided hash in memory with the existing hash. So since I used `.merge` here `@options_v4` is not modified outside of the test!